File: app/src/main/java/com/kinwatt/powermeter/ui/MapActivity.java
package com.kinwatt.powermeter.ui;

import android.graphics.Color;
import android.os.Bundle;
import android.view.WindowManager;
import android.widget.Button;

import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.JointType;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.Polyline;
import com.google.android.gms.maps.model.PolylineOptions;
import com.google.android.gms.maps.model.RoundCap;
import com.kinwatt.powermeter.R;
import com.kinwatt.powermeter.data.Position;
import com.kinwatt.powermeter.ui.widget.ChronometerView;
import com.kinwatt.powermeter.ui.widget.NumberView;

public class MapActivity extends ActivityBase implements OnMapReadyCallback {

    private GoogleMap mMap;
    private PolylineOptions options;
    private Polyline line;

    private MapController controller;

    private Button buttonStart, buttonStop;

    private NumberView power;
    private NumberView speed;
    private ChronometerView duration;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_map);

        buttonStart = findViewById(R.id.button_start);
        buttonStop = findViewById(R.id.button_stop);

        buttonStop.setEnabled(false);

        duration = findViewById(R.id.duration);
        speed = findViewById(R.id.speed);
        power = findViewById(R.id.power);

        speed.setUnits("km/h");
        power.setUnits("W");

        buttonStart.setOnClickListener(v -> {
            buttonStart.setEnabled(false);

            duration.restart();
            controller.start();

            getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);

            buttonStop.setEnabled(true);
        });

        buttonStop.setOnClickListener(v -> {
            buttonStop.setEnabled(false);

            getWindow().clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);

            duration.stop();
            controller.stop();

            buttonStart.setEnabled(true);
        });

        SupportMapFragment mapFragment = (SupportMapFragment) getSupportFragmentManager()
                .findFragmentById(R.id.map);
        mapFragment.getMapAsync(this);

        controller = new MapController(this);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        controller.stop();
    }

    public void onMapReady(GoogleMap googleMap) {
        mMap = googleMap;

        mMap.setMapType(GoogleMap.MAP_TYPE_HYBRID);

        LatLng malaga = new LatLng(36.7161622, -4.4233658);
        mMap.moveCamera(CameraUpdateFactory.newLatLng(malaga));
        mMap.moveCamera(CameraUpdateFactory.zoomTo(15));

        options = new PolylineOptions()
                .width(15)
                .color(Color.BLUE)
                .geodesic(true)
                .jointType(JointType.ROUND)
                .startCap(new RoundCap())
                .endCap(new RoundCap());
    }

    public void updateMap(Position position) {
        if (mMap != null) {
            mMap.animateCamera(CameraUpdateFactory.newLatLng(new LatLng(position.getLatitude(), position.getLongitude())));

            LatLng point = new LatLng(position.getLatitude(), position.getLongitude());

            if (!options.getPoints().contains(point)) {
                options.add(point);
            }

            Polyline newLine = mMap.addPolyline(options);
            if (line != null) {
                line.remove();
            }
            line = newLine;
        }
    }

    public void stopTimer() {
        duration.stop();
        buttonStop.setEnabled(false);
        buttonStart.setEnabled(true);
    }

    public void setSpeed(float speed) {
        this.speed.setValue(speed);
    }

    public void setPower(float power) {
        this.power.setValue(power);
    }

    public void clearMap() {
        if (mMap != null) {
            mMap.clear();
            options.getPoints().clear();
        }
    }
}

File: app/src/main/java/com/kinwatt/powermeter/ui/activities/MapActivity.kt
package com.kinwatt.powermeter.ui.activities

import android.os.Bundle
import android.view.WindowManager

import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.JointType
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.Polyline
import com.google.android.gms.maps.model.PolylineOptions
import com.google.android.gms.maps.model.RoundCap
import com.kinwatt.powermeter.R
import com.kinwatt.powermeter.data.Position
import com.kinwatt.powermeter.databinding.ActivityMapBinding
import com.kinwatt.powermeter.ui.MapController

class MapActivity : ActivityBase(), OnMapReadyCallback {

    private lateinit var binding: ActivityMapBinding

    private lateinit var options: PolylineOptions
    private var mMap: GoogleMap? = null
    private var line: Polyline? = null

    private lateinit var controller: MapController

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityMapBinding.inflate(layoutInflater)
        val view = binding.root
        setContentView(view)

        binding.buttonStop.isEnabled = false

        binding.speed.units = "km/h"
        binding.power.units = "W"

        binding.buttonStart.setOnClickListener { v ->
            binding.buttonStart.isEnabled = false

            binding.duration.restart()
            controller.start()

            window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)

            binding.buttonStop.isEnabled = true
        }

        binding.buttonStop.setOnClickListener { v ->
            binding.buttonStop.isEnabled = false

            window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)

            binding.duration.stop()
            controller.stop()

            binding.buttonStart.isEnabled = true
        }

        val mapFragment = supportFragmentManager
            .findFragmentById(R.id.map) as SupportMapFragment
        mapFragment.getMapAsync(this)

        controller = MapController(this)
    }

    override fun onDestroy() {
        super.onDestroy()
        controller.stop()
    }

    override fun onMapReady(googleMap: GoogleMap) {
        mMap = googleMap

        //mMap.setMapType(GoogleMap.MAP_TYPE_HYBRID);

        val malaga = LatLng(36.7161622, -4.4233658)
        mMap!!.moveCamera(CameraUpdateFactory.newLatLng(malaga))
        mMap!!.moveCamera(CameraUpdateFactory.zoomTo(15f))

        options = PolylineOptions()
            .width(15f)
            .color(resources.getColor(R.color.colorPrimary))
            .geodesic(true)
            .jointType(JointType.ROUND)
            .startCap(RoundCap())
            .endCap(RoundCap())
    }

    fun updateMap(position: Position) {
        if (mMap != null) {
            mMap!!.animateCamera(CameraUpdateFactory.newLatLng(LatLng(position.latitude, position.longitude)))

            val point = LatLng(position.latitude, position.longitude)

            if (!options.points.contains(point)) {
                options.add(point)
            }

            val newLine = mMap!!.addPolyline(options)
            if (line != null) {
                line!!.remove()
            }
            line = newLine
        }
    }

    fun stopTimer() {
        binding.duration.stop()
        binding.buttonStop.isEnabled = false
        binding.buttonStart.isEnabled = true
    }

    fun setSpeed(speed: Float) {
        binding.speed.value = speed.toDouble()
    }

    fun setPower(power: Float) {
        binding.power.value = power.toDouble()
    }

    fun clearMap() {
        if (mMap != null) {
            mMap!!.clear()
            options.points.clear()
        }
    }
}
